### ISSUE TITLE
[Bug]: Style tags disappear from Head during Error/Catch Boundary

### ISSUE THREAD

**Issue by hgeldenhuys** [remix-run/remix#1136]
### Which Remix packages are impacted?

- `remix` (Remix core)
- `@remix-run/react`

### What version of Remix are you using?

1.1.1

### What version of Node are you using? Minimum supported version is 14.

16.13.0

### Steps to Reproduce

Inject style tags into header in `entry.server.tsx` during markup generation. Example:

```typescript
let markup = renderToString(
    <RemixServer context={remixContext} url={request.url} />
  );

  const html = markup.replace(
    "<head>",
    `<head>${getSSRStyles(markup, server)}`
  );

  responseHeaders.set("Content-Type", "text/html");

  return new Response("<!DOCTYPE html>" + html, {
    status: responseStatusCode,
    headers: responseHeaders,
  });
```

### Expected Behavior

The style tags should stay intact during 404, or at least after 404, refetch the HTML from server.entry. 

I'm guessing because Remix only generates links an meta tags in head dynamically, things like style and base will get overridden when Error boundary generates a new document. This is unfortunate, as initial load will contain Style tags that prevent the page from loading janky and will be lost during errors.

This prevents us from using major UI libraries like Material UI and Mantine, which generate dynamic styles based on frameworks like emotion. The CatchBoundary/Error should re-render from server-side, not client-side (rehydrate head tag from server.entry)

### Actual Behavior

Error/Catch boundaries removes everything in Head tag, not just the dynamically generated head-elements provided by the route, meaning style tags get removed, that are "global" styles.

**Comment by pixelstew** [comment]
I would also add in that script tags from the end of the body are removed too so remix context gets lost other among other things.

**Comment by pixelstew** [comment]
So far I have ascertained a few things relating to this issue.

- __IMPORTANT__ Landing directly on a 404 does not reproduce this issue, `entry.server` runs and produces expected results.
- Clicking a link in app to a non existent route loads catchBoundary without running `entry.sever` and thus styles generated in this step are not regenerated.
- As well as losing any user generated style tags from the head the Remix Context does not get loaded in catchBoundary  and then when user clicks 'back' it is missing from the previous (genuine) route too. Instead there is an empty `<script></script>` where context is usually generated.

**Comment by TheRealFlyingCoder** [comment]
Just to add you can't hit the "back" button in this scenario either, it will throw another error around appendingChild as the loader will run but have no context anymore (I assume)

![image](https://user-images.githubusercontent.com/58131578/148016301-6e9d4aa6-1294-4610-b1cb-3613180457ac.png)

**Comment by akomm** [comment]
I'm using myself tailwindcss, but `mui` library uses emotion, so everything from mui breaks.

**Comment by akomm** [comment]
Because the project I worked on was still in dev, I ignored this for a while. Now I'v checked back shortly and figured out the issue is kind of trivial and obvious:

The way how typical/default `root.tsx` is implemented, it re-renders the whole `Document`, including `body` and `head` on error and catch boundaries. In remix we have access to the whole document, which is nice, but on root error/catch boundary, we rerender something like that:

```ts
// ...
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width,initial-scale=1" />
        {title ? <title>{title}</title> : null}
        <Meta />
        <Links />
      </head>
// ...
```

Obviously this re-renders the head without the styles injected by `emotion` or similar lib. The styles were inserved probably directly via dom...

So to solve the issue, we need to either:
1. re-inject the styles using lib api if available
2. Create a layout (e. G. __App) route beneath `root.tsx` with its own error/catch boundary. This way everything will fall back  to its error/catch boundaries and `Document` one level up will not re-render. Still, when errors happen in this layout route, it will fall back to `root.tsx`

**Comment by akomm** [comment]
Well, the issue is that the top-most (root.tsx) error & catch boundary - if default unchanged - re-renders the whole document, including `head`. The JSX does not including the styles injected directly via dom by for example emotion. 

I think it could be partially worked around using some layout root with error and catch boundaries one level underneath the root.tsx. Still, if there is a problem in this layout route, it will fall back to root.tsx, but at least not for every error and you are not required to unnaturally have boundaries all over the place.

Alternative, I did some research, would be to re-inject all the styles in effect. However the used sheet from material UI at first glance does not seem to give access for the underlying sheet instance (I did only a glance check). Also that cause visual flickr on error when the styles are re-injected in effect.

**Comment by kevinbailey25** [comment]
I'm having a very similar issues (although using styled-components). Any catch/error boundaries that I create in the `root.tsx` required me to re-render the whole document. 

As @akomm said, it would be great to have some "layout" middle piece that only re-renders a piece in the body and not the whole document.

**Comment by kevinbailey25** [comment]
Any idea of how to resolve this? This is one of the last pieces preventing us from converting our apps at work to Remix.

**Comment by akomm** [comment]
@kevinbailey25 

Add [pathless layout route](https://remix.run/docs/en/v1/guides/routing#pathless-layout-routes)
Add [ErrorBoundary](https://remix.run/docs/en/v1/api/conventions#errorboundary) and [CatchBoundary](https://remix.run/docs/en/v1/api/conventions#errorboundary) in this pathless-layout-routes

Add the pathless route directly descending your `root` route.

Now the issue should only appear, if you have an error in your pathless route, but not any route beneath it.

**Comment by jpod32** [comment]
I'm experiencing a similar issue, and I was able to resolve component errors by doing what @akomm mentioned. However, even with using a CatchBoundary in the pathless route, 404s are still being unhandled, I'm assuming since they are not originating in the pathless route, any way to handle that?

**Comment by akomm** [comment]
@jpod32 I just tested manually throwing an 404 response and the catch boundary works in a regular route. I don't have yet the possibility to make a test on a pathless route. If you want me to check into it, you could create a reproduction sandbox. It is possible that catch boundary does not work when the routing fails, hence my manually thrown 404 response is not the same. Or it might be only issue with pathless routes. Or maybe you'v overlooked something?

**Comment by jpod32** [comment]
My issue is with 404s that aren't thrown manually, which can just be reproduced by entering an invalid url. From my attempts, this doesn't go to the catch boundary inside the pathless route.

**Comment by oscarnewman** [comment]
I'm also experiencing this, and can second @jpod32's experience. I tried the pathless route solution and had the same issue for 404s from invalid URLs.

Has anyone else found a workaround to this yet?

**Comment by kevinbailey25** [comment]
TL;DR: talked to Ryan Florence about it... so he at least knows it's a pain point for those of us trying to migrate to Remix, but I have no perfect solution.

I lucked out at Remix Conf to have Ryan Florence sit at my table for lunch during the workshop. He asked everyone what's one thing that you wished Remix had and I brought up a lot of the styling pains I've had while trying to migrate.

Remix really wants to give you full control of your network tab. Hence why they want you to set the url for your css in the `links` function. Which I totally get. If I could go back in time when we started this project at work, I would have picked tailwind, but at the time we picked styled-components. I think a lot of developers are used to writing css that sits along side their components and magically makes it into their app. My angular skills are a bit rusty now, but I think even with that framework you would specify a css file for each component if needed. 

Then if you jumped into styled-components... it was... write your css that goes with that component. 

Things get more complicated when you've written a component library that's a separate package that expects styled-components to just work. 

I get why Remix went the way they did, but we can't re-write the whole app to get away from styled-components... I just can't convince the higher ups right now that it's worth it. 

Hopefully they can help get some happy paths for us using css-in-js libs. :/

**Comment by akomm** [comment]
@kevinbailey25 you will also have this issue when not directly using css-in-js libraries, like me. Using tailwind, but a third party library uses css-in-js.

**Comment by absamo** [comment]
Any progress on this bug, it is the one thing preventing us to use Remix with Mantine. Please give us feedback so at least we understand why fixing this issue is not easy. Thank you for your awesome work!

**Comment by akomm** [comment]
@absamo its not exactly a bug, since it behaves as it should, the whole document is re-rendered and anything injected directly into the DOM, like normally for react, disappears.

But I agree this is an important to solve issue. Some statement would be nice, even if its "we can't for this reason" or "it would mean this and that". Maybe someone get an idea based on this how to still make it better, if a 100% perfect solution is not possible.

**Comment by kwiat1990** [comment]
> Any progress on this bug, it is the one thing preventing us to use Remix with Mantine. Please give us feedback so at least we understand why fixing this issue is not easy. Thank you for your awesome work!

It‘s the same for me. I have tried Remix, loved it immediately for its data fetching model and then I’ve noticed the bug. It was heartbreaking experience 😉.

**Comment by correiarmjoao** [comment]
Had the same problem when trying to use Mantine. I was able to work around it by reapplying the styles on the client using emotion cache, it should also work for other libraries that use emotion but will probably need another context for server styles.

Made a [small repo with an example](https://github.com/correiarmjoao/remix-with-mantine), let me know if it helps @kwiat1990 @absamo

**Comment by kwiat1990** [comment]
> Had the same problem when trying to use Mantine. I was able to work around it by reapplying the styles on the client using emotion cache, it should also work for other libraries that use emotion but will probably need another context for server styles.
> 
> Made a [small repo with an example](https://github.com/correiarmjoao/remix-with-mantine), let me know if it helps @kwiat1990 @absamo

Wow, thanks for the repo. I forked it and I will definitely try it after my vacations!

**Comment by akomm** [comment]
Yes, basically this:
> Alternative, I did some research, would be to re-inject all the styles in effect. However the material UI style engine at first glance does not seem to give access for the underlying emotion sheet instance (I did only a glance check). Also, re-injecting as effect might cause flickr after initial boundary render

But sometimes you use it via other libraries, which don't expose the instances of cache they'v created.

**Comment by brandiqa** [comment]
> Had the same problem when trying to use Mantine. I was able to work around it by reapplying the styles on the client using emotion cache, it should also work for other libraries that use emotion but will probably need another context for server styles.
> 
> Made a [small repo with an example](https://github.com/correiarmjoao/remix-with-mantine), let me know if it helps @kwiat1990 @absamo

Thanks @correiarmjoao

I can confirm this works. Simply pay attention to the following files when making changes to your own project:

- [Document.tsx](https://github.com/correiarmjoao/remix-with-mantine/blob/main/app/components/Document.tsx)
- [context.tsx](https://github.com/correiarmjoao/remix-with-mantine/blob/main/app/context.tsx)
- [entry-client.tsx](https://github.com/correiarmjoao/remix-with-mantine/blob/main/app/entry.client.tsx)
-[entry-server.tsx](https://github.com/correiarmjoao/remix-with-mantine/blob/main/app/entry.server.tsx)

There is though the problem of Flash Of Unstyled Content in Mantine that occurs when navigating to a 404 page. It's not a big deal for me but something you should be aware of. Other supported frameworks(Gatsby, Next) and Chakra UI users are also facing the same issue.

**Comment by kwiat1990** [comment]
With Remix I didn’t try it yet. But in my Next project I needed to switch to yarn and downgrade Next to version `12.1.4` to get rid of FOUC.

**Comment by absamo** [comment]
> Had the same problem when trying to use Mantine. I was able to work around it by reapplying the styles on the client using emotion cache, it should also work for other libraries that use emotion but will probably need another context for server styles.
> 
> Made a [small repo with an example](https://github.com/correiarmjoao/remix-with-mantine), let me know if it helps @kwiat1990 @absamo

Awesome, nice work around. Thanks for sharing it.

**Comment by kwiat1990** [comment]
@correiarmjoao I've tried it and it seems to be working but I've found odd behaviour: if some route doesn't exist all Mantine's styles would be correctly applied but not those for body. In my case`<body>` gets default margin of `8px`, which normally is overridden by Mantines global reset.

**Comment by liho00** [comment]
same issue here +1

**Comment by kwiat1990** [comment]
@correiarmjoao do you have an idea how it can be done for the new Mantine 5.0? There was change to cache and this solution doesn't work anymore. I have also played with it a little bit but I can't really get it working.

**Comment by dfyz011** [comment]
I have the same problem with styled-components. I like ideas of @akomm and @hgeldenhuys to have mechanism that allows us to copy head to newly rendered document or re-render only body.

**Comment by correiarmjoao** [comment]
@kwiat1990 I had the same problem today, I updated the [repo](https://github.com/correiarmjoao/remix-with-mantine) to work with version 5

**Comment by muco-rolle** [comment]
Hello everyone,

I'm facing the same issue with Chakra UI was there anyone who was able to find the fix with Chakra UI?

**Comment by kevinbailey25** [comment]
@muco-rolle 

I have A solution. It's not ideal but it seems to work pretty well. (my solution is using styled-components, but I assume it should work with any css-in-js lib that adds styles to the <head> of the page)

`routes/root.tsx` <-- your root should have your <head> and <body> tags. You also shouldn't have any `loader` or `action` in this file. It also doesn't hurt to set `export const unstable_shouldReload = () => false`. The goal here is to make our root never re-render. Re-rendering is what causes all the styles to be removed (from my understanding). My root is pretty bare bones, it doesn't have much in here. 

Now for the magic part. [Pathless Layout Routes](https://remix.run/docs/en/v1/guides/routing#pathless-layout-routes)

Create a new pathless layout route that will be like your new "root". I call mine `__main` or even `__root`. Do what makes sense to you.
`routes/__main` <-- folder containing all your normal route modules
`routes/__main.tsx` <-- this is almost like your new root route. My route component, it literally just a `<Outlet />`. 

The other important parts in your `routes/__main.tsx` you should have a CatchBoundary and ErrorBoundary. 

I also like to add a catch all route that I can force a 404. So my CatchBoundary in my __main and catch it. `routes/__main/$.tsx`. 

I hope this all makes sense, but this is what I've been doing and so far I haven't had any issues. I'm hoping to make a sample project showing this, but I haven't gotten around to it yet.

**Comment by muco-rolle** [comment]
Thanks, @kevinbailey25 this definitely helped me understand a lot of things.

Following your descriptions, I was able to make a [working example](https://github.com/muco-rolle/remix-chakra-ui) I would appreciate very much your review

**Comment by muco-rolle** [comment]
I couldn't ask for a better explanation and guidance than that. Thanks @kevinbailey25

**Comment by jgentes** [comment]
I'd like to reiterate the solution that @kevinbailey25 provided with what I discovered:

1. Adding `ErrorBoundary` and `CatchBoundary` to `root` will not work if you use any form of styling (~ everyone?)

So the fix is to create a `__boundary` which is used to export these functions using a file structure of:

```
app
├── root.tsx
└── routes
    ├── __boundary.tsx 
    ├── __boundary
    │   ├── customers.tsx
    │   └── .. other pages within your layout
    └── $.tsx // to catch 404 errors
```

2. Root must contain your page layout with the `<Outlet />` component, and the `__boundary` should only export `<Outlet />`. If any other components are placed inside `__boundary` they will render twice - once when `root` is rendered, then again when picked up by the pathless (__) route.

With this approach everything works as expected, although a) this should be clarified on the [Routing page](https://remix.run/docs/en/v1/guides/routing) or better yet it would be great if b) Remix could create a virtual boundary within `root` so that the `ErrorBoundary` and `CatchBoundary` could be placed there instead of needing this pathless workaround.

**Comment by kevinbailey25** [comment]
@jgentes thanks for the reiteration of my solution, that's a much easier explanation of this work around than what I said. I do want to point out 2 things

1. I believe this work around is only needed for projects that use any kind of CSS-in-JS library for styling (aka any javascript styling library that injects `<style>` tags into the `<head>` or `<body>`).  If you're using css files (whether plain, or generated from things like Tailwind, PostCSS, etc) and export them via the [Route Module links exports](https://remix.run/docs/en/v1/guides/styling) you won't need to do this work around. I've got several projects that use Tailwind or css files that work perfectly fine.

2. Make sure your `$.tsx` route (to catch 404 errors) is inside your `__boundary` folder. Otherwise it'll try to render the 404 page using the `CatchBoundary` of your `root.tsx` instead of the one in your `__boundary.tsx`. 

 ```
 app
 ├── root.tsx
 └── routes
     ├── __boundary.tsx 
     └── __boundary
         ├── customers.tsx
         ├── .. other pages within your layout  
         └── $.tsx // to catch 404 errors
 ```

Hope this helps people!

**Comment by jgentes** [comment]
Thanks @kevinbailey25 I've updated my answer with the $.tsx moved into __boundary

**Comment by kevinbailey25** [comment]
Just want to throw out a `GOTCHA` that kicked my butt with all of this last week. 

TL;DR: stay on React 17 for now. 

All of my work arounds I've listed previously is to prevent the `<head>` tag from being rehydrated/replaced because you'll lose all of your styles. 

Insert React 18.2 and anyone that has a browser plugin that inserts anything into the page:
https://github.com/facebook/react/issues/24430

The way React 18.2 currently hydrates the page, if anything that was generated from the server doesn't match what React wants to hydrate, React throws everything out and re-renders the document. Which is remix includes throwing out the head and replacing it, which throws out styles. 

My team worked hard making sure that what gets server rendered is the same as what gets hydrated. Even with complicated things like localization, dates in different time zones, etc. Then we discovered if you have any browser extension that injects anything into the page (LastPass for example) this causes React 18.2 to re-render everything (because it will never match what we server rendered, and we can't stop users from using plugins). Thus losing your styles.

We feel very confident that in the future, either React will have a better solution for this, or even your CSS-in-JS libraries will have a solution to this. So we're sticking with React 17 for now, we are considering going away from CSS-in-JS but we've got multiple years of component libraries and other products that are all built with CSS-in-JS, so not an easy switch by any means.

Hope this helps everyone in the meantime.

**Comment by muco-rolle** [comment]
> I'd like to reiterate the solution that @kevinbailey25 provided with what I discovered:
> 
> 1. Adding `ErrorBoundary` and `CatchBoundary` to `root` will not work if you use any form of styling (~ everyone?)
> 
> So the fix is to create a `__boundary` which is used to export these functions using a file structure of:
> 
> ```
> app
> ├── root.tsx
> └── routes
>     ├── $.tsx // to catch 404 errors
>     ├── __boundary.tsx 
>     ├── __boundary
>     │   ├── customers.tsx
>     │   └── .. other pages within your layout
> ```
> 
> 2. Root must contain your page layout with the `<Outlet />` component, and the `__boundary` should only export `<Outlet />`. If any other components are placed inside `__boundary` they will render twice - once when `root` is rendered, then again when picked up by the pathless (__) route.
> 
> With this approach everything works as expected, although a) this should be clarified on the [Routing page](https://remix.run/docs/en/v1/guides/routing) or better yet it would be great if b) Remix could create a virtual boundary within `root` so that the `ErrorBoundary` and `CatchBoundary` could be placed there instead of needing this pathless workaround.

Can you please share an example repo where you implemented this?

**Comment by kevinbailey25** [comment]
Here's an example repo that I've tossed together using `styled-components`.
https://github.com/kevinbailey25/remix-styled-components-example

**Comment by jgentes** [comment]
> Can you please share an example repo where you implemented this?

[https://github.com/jgentes/mixpoint](https://github.com/jgentes/mixpoint)

**Comment by jca41** [comment]
I'm also seeing a flash of unstyled content with normal css.

Currently i'm developing an app and i have a `<Link />` to a bunch of missing pages. I know this is not the "normal setup" as you're not really meant to have missing links within your app!

Repro:

- Open in preview mode
- If you disable cache in devTools you might be able to see the white flash for a split second.

https://stackblitz.com/edit/node-5plrzf?file=app%2Froot.tsx

**Comment by kevinbailey25** [comment]
> I'm also seeing a flash of unstyled content with normal css.
> 
> Currently i'm developing an app and i have a `<Link />` to a bunch of missing pages. I know this is not the "normal setup" as you're not really meant to have missing links within your app!
> 
> Repro:
> 
> * Open in preview mode
> * If you disable cache in devTools you might be able to see the white flash for a split second.
> 
> https://stackblitz.com/edit/node-5plrzf?file=app%2Froot.tsx

This is pretty standard to how the web works in general. [This is not unique to remix](https://en.wikipedia.org/wiki/Flash_of_unstyled_content). With how you've set up your `root.tsx` you can simulate the same effect using 2 static html pages... one for your root and one for your 404 page (catch boundary). If you are telling your browser to disable the cache, then it will have to load the css file for each page.

**Comment by jca41** [comment]
> > I'm also seeing a flash of unstyled content with normal css.
> > Currently i'm developing an app and i have a `<Link />` to a bunch of missing pages. I know this is not the "normal setup" as you're not really meant to have missing links within your app!
> > Repro:
> > 
> > * Open in preview mode
> > * If you disable cache in devTools you might be able to see the white flash for a split second.
> > 
> > https://stackblitz.com/edit/node-5plrzf?file=app%2Froot.tsx
> 
> This is pretty standard to how the web works in general. [This is not unique to remix](https://en.wikipedia.org/wiki/Flash_of_unstyled_content). With how you've set up your `root.tsx` you can simulate the same effect using 2 static html pages... one for your root and one for your 404 page (catch boundary). If you are telling your browser to disable the cache, then it will have to load the css file for each page.

Should it not hydrate the new page correctly? The css file is present on both "pages".
The cache off suggestion was just to make the issue occur frequently.

Does this mean it's probably better to keep a pathless layout root below `root` to avoid the removal and re-insertion of the css file?

**Comment by muco-rolle** [comment]
@kevinbailey25 have you tried solutions like Vanilla Extract? would you recommend it if it's a project that is in the early stage where migrating to it would not hurt?

**Comment by machour** [comment]
@muco-rolle should work fine, we have an example in the examples repo. 
Not the ideal implementation, but it's done like tailwind so you shouldn't see the problem raised in this issue.

**Comment by kevinbailey25** [comment]
> @kevinbailey25 have you tried solutions like Vanilla Extract? would you recommend it if it's a project that is in the early stage where migrating to it would not hurt?

I have not personally used Vanilla Extract. Based off of everything I've seen and read, I'd recommend it. Just about any css framework/tool that gives you a standard css file that you can link to is recommended by me. As @machour  mentioned above, there is an example for remix as well.

**Comment by JAD3N** [comment]
This issue is probably related to #4175

**Comment by muco-rolle** [comment]
@kevinbailey25 can this https://github.com/Xiphe/remix-island fix this issue?

**Comment by kevinbailey25** [comment]
I saw this the other day, in theory it could. Although the library even mentions at the end of their [readme](https://github.com/Xiphe/remix-island#pitfallsnotes) that it may have unexpected results with libraries that inject things into the head like `styled-components` and such. 

At my work, we've decided to move away from CSS-in-JS libraries and move towards solutions that give us css files to link on the page and not have the extra overhead of runtime css.

**Comment by muco-rolle** [comment]
Thanks for your feedback @kevinbailey25. 

What tools are you using now?

**Comment by kevinbailey25** [comment]
> Thanks for your feedback @kevinbailey25.
> 
> What tools are you using now?

We have a component/design library that we've fully removed `styled-components` from in favor of plain css. Our library ships with a single css file for the whole thing. (We may split that up into smaller css files the future if we find necessary). 

All of our products that consume that component library, are configured to use Tailwind.

**Comment by danramos-evoiq** [comment]
This still seems to be an issue, any update?

**Comment by markdalgleish** [comment]
Thanks for raising this issue, and thanks everyone for helping out with all the different fixes and workarounds.

Since this affects multiple libraries and the workaround for each one is going to be different, I'm going to close this overarching issue just so it's not open indefinitely. If there are more specific issues related to a single CSS-in-JS library, UI framework etc., please feel free to open a separate issue.

Transparently supporting any library that injects global side-effects into the document would require a bigger change to Remix's architecture. This is something we'll keep in mind moving forwards but we don't have any concrete plans to address this at the moment.

In the meantime I'd encourage everyone to help contribute to the [Remix examples repo](https://github.com/remix-run/examples) with any fixes or workarounds for specific libraries. Note that we already have a working [Emotion example](https://github.com/remix-run/examples/tree/main/emotion) and I've just opened a [PR that fixes the Mantine example](https://github.com/remix-run/examples/pull/309) based on the comments in this issue (thanks @correiarmjoao!).

**Comment by thebeyondr** [comment]
My styles were disappearing when my error boundary got triggered. I'm using the Vite version with TailwindCSS. 

I fixed my issue here in case this helps anyone: https://github.com/remix-run/remix/discussions/6356?sort=top#discussioncomment-8128826

**Comment by kiliman** [comment]
@NTag do you have a simple repo?

**Comment by NTag** [comment]
Now I do, just here: https://github.com/NTag/remix-demo-vite-errors-css

See the demo:
![CleanShot 2024-03-19 at 14 52 58](https://github.com/remix-run/remix/assets/906276/dc8c8cf4-f6d4-47e2-858a-33dc6753a0d7)

Also, when creating the repo, I noted that the issue only appears after a first navigation. If you directly go (in my repo) to the `/about` page, refresh, click the broken link, and press back, everything is fine, the style stays.

However if you first start with the home page (like in the viceo), then navigate to the `/about` page and then click on the broken link, and back, this is when the styles completely disappears.

Hope it helps!

**Comment by seanmcquaid** [comment]
@NTag Did you ever find a solution to this problem?? I have been seeing the same thing even after implementing a Layout component exported in root.tsx.

**Comment by NTag** [comment]
No solution yet no…

**Comment by mosino** [comment]
I found a (bit of a hacky) workaround:

root.tsx
```
const error = useRouteError();
const navigate = useNavigate()

useEffect(() => {
    if(isRouteErrorResponse(error)) {
       window.location.href = "./errorPage"
    }
}, [error, navigate]);
```

And of course, put an ErrorPage.tsx in the routes.

I agree that this is a major bug

**Comment by mhmdjaw** [comment]
This is still happening and things really get messed up when I'm using animation and UI libraries like framer-motion and react-konva.

**Comment by lifeiscontent** [comment]
@mhmdjaw @mosino @NTag this is how you fix the issue: https://github.com/NTag/remix-demo-vite-errors-css/pull/1

**Comment by mhmdjaw** [comment]
> @mhmdjaw @mosino @NTag this is how you fix the issue: [NTag/remix-demo-vite-errors-css#1](https://github.com/NTag/remix-demo-vite-errors-css/pull/1)

I'm already doing that. It still happens.

**Comment by mosino** [comment]
> > @mhmdjaw @mosino @NTag this is how you fix the issue: [NTag/remix-demo-vite-errors-css#1](https://github.com/NTag/remix-demo-vite-errors-css/pull/1)
> 
> I'm already doing that. It still happens.

same here

**Comment by lifeiscontent** [comment]
Hmm, the repro that @NTag put together seems to be fixed, is there something I missed? or is what you're talking about not outlined in the repro @mhmdjaw @mosino ?

**Comment by mhmdjaw** [comment]
@lifeiscontent I believe it still happens although in my case (I'm using CSS modules) not all styles are affected. Whenever I navigate to the 404 ErrorBoundary component and navigate back, some styles get messed up. It seems to be happening to the same styles when I test again, so although it seems random it's probably not. I don't know why it happens to certain styles only, but I can try to reproduce with a simple setup if you wish.

**Comment by lifeiscontent** [comment]
@mhmdjaw yeah, I'd love to try and help figure it out, if you can put a repro together, I'll see what I can come up with.

**Comment by mhmdjaw** [comment]
> @mhmdjaw yeah, I'd love to try and help figure it out, if you can put a repro together, I'll see what I can come up with.

I actually did include a very simple repro, and an edit to my comment a few hours ago. Tell me if you need anything else.

**Comment by lifeiscontent** [comment]
@mhmdjaw sorry, just saw your comment, where can I find the code for the repro? I don't see it.

**Comment by lifeiscontent** [comment]
@mhmdjaw here's how you fix it with ssr: false

https://github.com/NTag/remix-demo-vite-errors-css/pull/1/commits/6d268086a3c456eb1ae038ad5648cca6a93f2c32

**Comment by mhmdjaw** [comment]
@lifeiscontent Unfortunately it still didn't solve the problem. If you're still curious about solving it, here's the [repo](https://github.com/mhmdjaw/remix-spa-styles-bug). My bad I forgot to push last time.

**Comment by lifeiscontent** [comment]
@mhmdjaw fixed with splat routes: https://github.com/mhmdjaw/remix-spa-styles-bug/pull/1

**Comment by lifeiscontent** [comment]
@mhmdjaw did that work for you?

**Comment by mhmdjaw** [comment]
@lifeiscontent Yup. That's the right workaround but it still doesn't solve the issue of losing the styles when navigating back from the ErrorBoundary, since this solution avoids it altogether.

**Comment by lifeiscontent** [comment]
@mhmdjaw yeah, another idea would be to add reloadDocument on the link, but that seems heavy handed.

**Comment by mhmdjaw** [comment]
@lifeiscontent yeah, what I ended up doing in my actual project was using a normal anchor instead of the remix Link which reloads the app at the home page. Switching to browser navigation is definitely not ideal, but it's just an error page so it's not a big deal. Otherwise your workaround with splat route is totally fine when the error is 404 not found.

**Comment by rhyek** [comment]
> @lifeiscontent yeah, what I ended up doing in my actual project was using a normal anchor instead of the remix Link which reloads the app at the home page. Switching to browser navigation is definitely not ideal, but it's just an error page so it's not a big deal. Otherwise your workaround with splat route is totally fine when the error is 404 not found.

This is a good enough solution for me.

**Comment by bakikucukcakiroglu** [comment]
I feel that [sonner toaster style issue](https://github.com/remix-run/remix/discussions/9318) and this is somehow related.
